packages/contracts/contracts/L1/ValidatorManager.sol

```diff
@@ -736,4 +765,13 @@ contract ValidatorManager is ISemver, IValidatorManager {
             _validatorInfo[validator].noSubmissionCount = 0;
         }
     }
+
+    /**
+     * @notice Allow restrictedOutputIndex.
+     *
+     * @param _outputIndex The output index to set the limit on.
+     */
+    function setRestrictedOutputIndex(uint256 _outputIndex) public onlyTrustedValidator {
+        restrictedOutputIndex = _outputIndex;
+    }
```

Comment:
How about making `restrictedOutputIndex` immutable and putting this into constructor? If trusted validator key is compromised, the attacker can delay the protocol, by submitting invalid output and setting this index continuously until the contract is upgraded to fix this. 

Setting it as immutable value can induce a situation that one output can still only be submitted by trusted validator even if we should revert MPT migration for some reasons, but still it would be better than the above situation. Hdyt?

Reply:
While restrictedOutputIndex supports the ability to restrict the submission of output, there is one consideration.  
In the event of a contingency, we may want to extend the MPT time into the future and also extend the restrictedOutputIndex into the future. 
This should be as reactive as possible, so I thought of a separate setter rather than a contract upgrade.

Reply:
What I mean was that we don't have to set `restrictedOutputIndex` again in the emergency of reverting the upgrade. Just letting the trusted validator exclusively submit in that index does not impact protocol much, and we can handle it by upgrading the contract later to update that parameter. Rather, I thought giving the permission of setting an exclusive output (any time, even he can set it multiple times) to trusted validator can be quite excessive and dangerous in the instance of hack.

But everything is tradeoff, and we can discuss and decide how to handle this.

Reply:
If you're concerned about changing `restrictedOutputIndex` multiple times, restricting it to be changed only once could be an option.

Reply:
Agreed, that can also be an option.

Reply:
Yes, this is a good discussion to be cautious about restrictedOutputIndex. 
Currently, restrictedOutputIndex has two functions
- allow submissions only for trusted validators for that output index
- disallow create challenge for that output index (trusted)
I think that if there is an urgent need to change restrictedOutputIndex, the second function above (unblocking create challenge) should also work with it. Thinking about this, it seems like the way to do it via upgrade might be problematic.

Translated with www.DeepL.com/Translator (free version)

Reply:
Yes, I get it
It would be much safer to take MPT_FIRST_OUTPUT_INDEX to immutable and upgrade it.
I'll apply the above.